hera_sim/simulate.py

```diff
@@ -434,4 +451,69 @@ def add_xtalk(self, model='gen_whitenoise_xtalk', bls=None, **kwargs):
                 vis=self.data.data_array[blt_ind, 0, :, pol_ind],
                 xtalk=xtalk
             )
+    
+    
+    def run_sim(self, sim_file=None, **sim_params):
+        """
+        Accept a dictionary or YAML file of simulation parameters and add in
+        all of the desired simulation components to the Simulator object.
+
+        Args:
+            sim_file (str): string providing a hook to a YAML file
+            
+            sim_params (dict): dictionary of simulation parameters.
+                it must take the form {model:params}, where model is
+                a string specifying the simulation component to be added
+                and params is a dictionary providing all the keyword
+                arguments and their desired values.
+
+            If both sim_file and sim_params are provided, then this function
+            will raise an AssertionError.
+        """
+
+        # keep track of which components don't use models
+        uses_no_model = []
+        for key, val in self.SIMULATION_COMPONENTS.items():
+            if 'model' not in inspect.signature(val).parameters:
+                uses_no_model.append(key)
+
+        assert sim_file is not None or len(sim_params) is not 0, \
+                'Either a hook to a simulation file or a dictionary of ' + \
+                'simulation parameters must be provided.'
+
+        assert sim_file is None or len(sim_params) is 0, \
+                'Either a simulation configuration file or a dictionary ' + \
+                'of simulation parameters may be passed, but not both. ' + \
+                'Please choose only one of the two to pass as an argument.'
+
+        # if a hook to a simulation file is provided, then read it in
+        if sim_file is not None:
+            with open(sim_file, 'r') as doc:
+                try:
+                    sim_params = yaml.load(doc.read(), Loader=yaml.FullLoader)
```

Comment:
This is going to be problematic for simulation components that require an interpolation object for the sky temperature model. I can't currently think of a *nice* way of handling these cases. (A straightforward way would be to check for the `Tsky_mdl` key in any of the parameter dictionaries, parse the value, and point to the interpolation object corresponding to it, but this requires the interpolation object to already exist somewhere.)

Reply:
Hmmm, what exactly is the problem here? Is the problem that it's hard to specify an interpolation object in a YAML file? If you have a dict of available interpolation objects defined somewhere inthe code, you could reference them by name in the YAML file? They definitely have to exist, of course. 

Or is it that you are envisioning that people may create their own interpolation objects outside of `hera_sim` and may want to use them?

Reply:
The problem here is indeed that it's hard to specify an interpolation object in a YAML file. Another difficulty is that (as far as I know) interpolation objects cannot be saved to disk. (If there is a way to do so that is machine-agnostic, then I'd be happy to be informed of such a way.)

I think that making a dict of available interpolation objects defined somewhere in the code is a quick and easy fix, but I don't think that it's a fix that's made with longevity in mind. I think that it would be better to allow people to make their own interpolation objects outside of `hera_sim` and use them; however, that seems much easier said than done.

There is one obvious way I can think of going about doing this. It would involve creating a helper function `_make_interp_object`, which accepts a hook to a `npz` file which has in it an array of `Tsky` values, as well as arrays of the LSTs and frequencies that the model is evaluated at. We could enforce a regular formatting to this: the frequency array must be saved with the key `'fqs'`; the LST array must be saved with the key `'lsts'`; and the sky temperature model must be saved with the key `'Tsky'`. The helper function would then generate and return an interpolation object using the arrays stored in the `npz` file. I imagine we could also include the ability to choose which type of interpolator to use (but limit it to types supported by `scipy.interpolate`), and allow for users to pass `'h1c'` or `'h2c'` (and so on) for shortcuts to the accepted standard sky temperature model for the given observing season.

Any thoughts on this?

Reply:
I think your idea is a good one -- though I reckon it's better as a class:

```
class Tsky:
    def __init__(self, datafile, interpolator='linear'):
        self.datafile = datafile
        self.interpolator = interpolator
        # np.load is lazy, so this is fast
        self._data = np.load(self.datafile)
    
    @property
    def freqs(self):
        return self._data['freqs']
    @property
    def tsky(self):
        return self._data['tsky']
    @property
    def lsts(self):
        return self._data['lsts']

    @cached_property
    def _interpolator(self):
       return scipy.interpolate(...)

    def __call__(self, freqs, lsts):
        return self._interpolator(freqs, lsts)
```

Then when loading from YAML you just need to use this class.
            

```

Reply:
Any suggestions of which module to put this class in?

Reply:
either a new module, or maybe `noise`? I'd go for a new module.

Reply:
If we put it in a new module, do we expect to add new things to the module other than the `Tsky` class? Any ideas for what to name the module?

Reply:
Another question: do we provide the option for users to choose any interpolator from the `scipy.interpolate` module, and require that they specify which with a string? Or do we restrict it to just a few types of interpolators (e.g. `RectBivariateSpline` and `interp1d`)?

Reply:
Honestly, I'd keep to a single one, and let the user pass kwargs to that interpolator (like kind='linear'). All the interpolators have different API's. I think it's up to you to choose the one that best suits this problem and just use it.

Reply:
So I've been thinking a bit about this, and the proposed implementation would break the current `hera_sim` API a bit. In particular, `diffuse_foreground` and `thermal_noise` require an interpolation object to be passed, rather than the re-sampled interpolation object. So the proposed method wouldn't jive with what's currently in place.

Here's what I'm thinking in order to solve this problem:
Instead of having a `__call__` method that returns the re-sampled sky temperature model, we implement `noise.resample_Tsky` as a member function of the `Tsky` class. Additionally, we keep important metadata in the `Tsky` class, in particular the units of class properties and the polarizations supported. In order to retrieve the interpolation object, we can implement a `get_interpolator` member function. 

Any thoughts on this? Also, do you have any thoughts on putting the `Tsky` class in the `utils` module?

Reply:
Another question/comment regarding your suggested format for the `Tsky` class:
What's the motivation behind using 
```
@property
def freqs(self):
    return self._data['freqs']
```
instead of `self.freqs = self._data['freqs']` in the `__init__` method? I understand that this makes it so that a `Tsky` object's frequencies (and so on) can't be changed, since it just pulls them from the loaded `npz` file; however, is this a feature that we want?

Reply:
I think maybe you mis-understand what `__call__` does? The class that I suggested would act like this:

```
tsky = Tsky("my_data_file.npz")
tsky_at_lsts_and_freqs = tsky(lsts, freqs)

# Can pass directly in as-is
diffuse_foregrounds(lsts, freqs, bl_vec, tsky)
```

i.e. the `__call__` method means you can call the instance itself. In this case, that call will do the resampling, just as a spline object would.

You could put it in `utils` for now -- also import it directly in `__init__` so that if we swap its module it won't hurt too many users.

Concerning the use of `@property`s, I do think we should disallow setting freqs/lsts by the user. If they are changed, the whole object becomes inconsistent. If you want to change them, you should create a new object from scratch. Also, this way the data is lazy-loaded.

Reply:
Ah, yes, I see what you mean.

 That said, I still think there's some value to having the `Tsky` object instead hold references to the interpolator objects. As a working example, the sky temperature model that's been used in `hera_sim` thus far (saved in `HERA_Tsky_vs_LST.npz`) has sky temperature arrays for two polarizations. If we construct the class so that the resampling method is accessed via `__call__`, then we would need to either have the user specify the polarization used when a `Tsky` object is instantiated, or we would need to enforce that the `npz` files hold data for only one polarization. (We could require that the polarization be specified in the `__call__` method, but that would mess with the API.)

Reply:
Hmm, that's what I think the `._interpolator` property is. It's a scipy interpolator, sitting in the class waiting to be used.

On the polarization thing: yeah I reckon you pass it at init-time, just like the "interpolation_kind" or whatever. You can imagine that constructing from a YAML is still easy: you specify the datafile and then the other options are strings.